go/pools/fast_pool.go

```diff
@@ -0,0 +1,627 @@
+			// a resource, causing a deadlock. We'll check occasionally to see
+			// if there is now capacity to create.
+			if p.State().Closed {
+				return nil, ErrClosed
+			}
+			wrapper, err := p.safeCreate(forUse)
+			if err == errNeedToQueue {
+				continue
+			} else if err != nil {
+				return nil, err
+			} else {
+				return wrapper.resource, nil
+			}
+		}
+	}
+}
+
+// Put will return a resource to the pool. For every successful Get,
+// a corresponding Put is required. If you no longer need a resource,
+// you will need to call Put(nil) instead of returning the closed resource.
+// The will eventually cause a new resource to be created in its place.
+func (p *FastPool) Put(resource Resource) {
+	p.Lock()
+
+	p.state.InUse--
+
+	if p.state.Closed || p.active() > p.state.Capacity {
+		// We're ether closed or shrinking.
+		p.Unlock()
+		if resource != nil {
+			resource.Close()
+		}
+		return
+	}
+
+	defer p.Unlock()
+
+	if resource == nil {
+		return
+	}
+
+	if p.state.InUse < 0 {
+		p.state.InUse++
+		panic(ErrPutBeforeGet)
+	}
+
+	w := resourceWrapper{resource: resource, timeUsed: time.Now()}
+	select {
+	case p.pool <- w:
+		p.state.InPool++
+	default:
+		// We don't have room in the pool.
+		p.state.InUse++
+		panic(ErrFull)
+	}
+}
+
+// SetCapacity changes the capacity of the pool.
+// You can use it to shrink or expand, but not beyond
+// the max capacity. If the change requires the pool
+// to be shrunk and `block` is true, SetCapacity waits
+// till the necessary number of resources are returned
+// to the pool.
+//
+// A SetCapacity of 0 is equivalent to closing the pool.
+func (p *FastPool) SetCapacity(capacity int, block bool) error {
+	p.Lock()
+	defer p.Unlock()
+
+	if p.state.Closed {
+		return ErrClosed
+	}
+
+	if capacity < 0 || capacity > cap(p.pool) {
+		return fmt.Errorf("capacity %d is out of range", capacity)
+	}
+
+	if capacity > 0 {
+		minActive := p.state.MinActive
+		if capacity < minActive {
+			return fmt.Errorf("minActive %v would now be higher than capacity %v", minActive, capacity)
+		}
+	}
+
+	if capacity == 0 {
+		p.state.Closed = true
+	}
+
+	isGrowing := capacity > p.state.Capacity
+	p.state.Capacity = capacity
+
+	if isGrowing {
+		p.state.Draining = false
+		return nil
+	}
+
+	p.state.Draining = true
+
+	if block {
+		p.shrink()
+	} else {
+		go p.withLock(p.shrink)
+	}
+
+	return nil
+}
+
+// shrink active resources until capacity it is not above the set capacity.
+// Requires the pool mutex to be locked when calling.
+func (p *FastPool) shrink() {
+	for {
+		remaining := p.active() - p.state.Capacity
+		if remaining <= 0 {
+			p.state.Draining = false
+			if p.state.Capacity == 0 {
+				close(p.pool)
+				p.state.Closed = true
+			}
+			return
+		}
+
+		// We can't remove InUse resources, so only target the pool.
+		// Collect the InUse resources lazily when they're returned.
+		p.Unlock()
+		select {
+		case wrapper := <-p.pool:
+			wrapper.resource.Close()
```

Comment:
If this panics, we would not re-lock. Not sure what the effects will be.

Reply:
Good catch. It caused a `unlock of unlocked mutex` in the caller.